### PR TITLE
[Configuration] Added new Alias field to the front end

### PR DESCRIPTION
**Pull request by ridz1208**
## Brief summary of changes
https://github.com/aces/Loris/issues/5539

### CODE REVIEW

File: modules/configuration/ajax/updateProject.php [hunk 1]
```diff
@@ -28,7 +28,7 @@
 $ProjectList = Utility::getProjectList();
 $projectName = $_POST['Name'] ?? '';
 // FIXME This field should be added to the front-end.
```
Comment (johnsaigle) [inline modules/configuration/ajax/updateProject.php]: this line should be deleted now that you've added this feature

File: modules/configuration/ajax/updateProject.php [hunk 2]
```diff
@@ -50,6 +50,7 @@
     // Update Project fields
     $project = \Project::getProjectFromID($projectID);
     $project->updateName($projectName);
+    $project->updateAlias($projectAlias);
```
Comment (johnsaigle) [inline modules/configuration/ajax/updateProject.php]: It would be good to add a back-end check to the front-end check that Alias is > 4.

An InvalidArgumentException should be thrown by the Project class when Alias > 4 characters.

Currently this is done only in `setAlias()`, but not in the constructor or in `updateAlias()`.

I'd suggest adding a `_validateAlias(): void` function that throws the Exception based on the `_projectAlias` variable and then calling that function in set, update, and the constructor.

Comment (johnsaigle) [reply modules/configuration/ajax/updateProject.php]: Come to think of it this would be a good use case for an Alias value object that can also be used by Site. 

We can make a separate issue for this and punt to the next release if you prefer.

Comment (ridz1208) [reply modules/configuration/ajax/updateProject.php]: Shouldnt we be then using the set functions and then updateProject, instead of separate update functions for each value ?

Comment (ridz1208) [reply modules/configuration/ajax/updateProject.php]: and for the alias object i would punt it because we dont have the same criteria for a psc alias, a mri alias and a project alias, we would need to standardise first

Comment (johnsaigle) [reply modules/configuration/ajax/updateProject.php]: Created #5565 for adding a value object.

Comment (johnsaigle) [reply modules/configuration/ajax/updateProject.php]: >Shouldnt we be then using the set functions and then updateProject, instead of separate update functions for each value ?

That would work too, just depends on how we want the class to work. I don't mind either way

File: modules/configuration/ajax/updateProject.php (hunk 1 above)
Comment (johnsaigle) [reply modules/configuration/ajax/updateProject.php]: I'll approve after you fix this; we can save the back-end check for later

File: modules/configuration/ajax/updateProject.php (hunk 2 above)
Comment (ridz1208) [reply modules/configuration/ajax/updateProject.php]: I'll add the validation still but we'll rethink the logic later

Comment (ridz1208) [reply modules/configuration/ajax/updateProject.php]: I made it pass through set which allowed me to remove the `??`

File: php/libraries/Project.class.inc [hunk 3]
```diff
@@ -227,13 +227,16 @@ class Project
     /**
      * Set this project's name
      *
-     * @param string $projectName This project's name
+     * @param ?string $projectName The new name. Uses class property when null.
      *
      * @return void
      */
-    public function updateName(string $projectName): void
+    public function updateName(?string $projectName = null): void
```
Comment (johnsaigle) [inline php/libraries/Project.class.inc]: Why did you add nullable here? Do we want the project name to be null?

File: php/libraries/Project.class.inc [hunk 4]
```diff
@@ -265,33 +268,18 @@ class Project
...
-     * Update the Alias field in the databse for this Project.
+     * Update the Alias field in the database for this Project.
      *
-     * @param ?string $alias The new alias. Uses class property when null.
+     * @param ?string $projectAlias The new alias. Uses class property when null.
      *
      * @return void
      */
-    public function updateAlias(?string $alias = null): void
+    public function updateAlias(?string $projectAlias = null): void
```
Comment (johnsaigle) [inline php/libraries/Project.class.inc]: Alias is non-nullable in the DB. Maybe we should keep this a string?
```suggestion
    public function updateAlias(string $projectAlias = ''): void
```

File: php/libraries/Project.class.inc [hunk 5]
```diff
@@ -341,10 +329,28 @@ class Project
      */
     public function updateRecruitmentTarget(?int $target = null): void
     {
-        $this->_updateField(
-            "recruitmentTarget",
-            $target ?? $this->_recruitmentTarget
-        );
+        if (isset($target)) {
+            $this->setRecruitmentTarget($target);
+        }
```
Comment (johnsaigle) [inline php/libraries/Project.class.inc]: I think this value will always be set, either to the value passed or `null`.

File: test/unittests/NDB_ConfigTest.php [hunk 6]
```diff
@@ -246,11 +246,13 @@ public function testGetProjectSettings()
         $info   = array(
                    'ProjectID'         => '999',
                    'Name'              => 'test',
+                   'Alias'             => 'TST',
```
Comment (johnsaigle) [inline test/unittests/NDB_ConfigTest.php]: Class constant would be nicer

File: php/libraries/Project.class.inc (hunk 3 above)
Comment (ridz1208) [reply php/libraries/Project.class.inc]: it defaults to the objects _projectName which is required by the singleton. the reason it is nullable here is if you do something like
```
setName("BLA");
updateName();
```
that would persist the name to BLA

File: php/libraries/Project.class.inc (hunk 4 above)
Comment (ridz1208) [reply php/libraries/Project.class.inc]: I had left that as is from your PR and adapted the other update functions to behave similarly.

since one validation is already being done in the set function, I would say we can add the non null check to the set function and the create new function maybe ?

File: php/libraries/Project.class.inc (hunk 5 above)
Comment (ridz1208) [reply php/libraries/Project.class.inc]: this should be doing the same thing as ur null coalesce, i dont think the behaviour changes.

it defaults to the classes' _recruitmentTarget if null is passed and if the classes in null it sets it to null i guess

File: test/unittests/NDB_ConfigTest.php (hunk 6 above)
Comment (ridz1208) [reply test/unittests/NDB_ConfigTest.php]: just for that value ?

File: php/libraries/Project.class.inc (hunk 3 above)
Comment (johnsaigle) [reply php/libraries/Project.class.inc]: Why not use an empty string instead of null?

Comment (ridz1208) [reply php/libraries/Project.class.inc]: not sure i understand, whats the advantage of an empty string ?

File: php/libraries/Project.class.inc (hunk 4 above)
Comment (johnsaigle) [reply php/libraries/Project.class.inc]: I think it makes sense to type hint as `string` so that it can't be null. Then we can use empty strings where we're using `null` and leave the empty() checks.

The types in the class should agree with the types in the DB or else we're inviting bugs.

File: php/libraries/Project.class.inc (hunk 5 above)
Comment (johnsaigle) [reply php/libraries/Project.class.inc]: Null coalesce handles `null` and not set. `isset()` only handles not set.

Comment (johnsaigle) [reply php/libraries/Project.class.inc]: When PHP 7.4 comes out we can use typed class properties so this won't be an issue. 😄

Comment (ridz1208) [reply php/libraries/Project.class.inc]: https://www.php.net/manual/en/function.isset.php

isset() only returns false when a value is NULL

Comment (johnsaigle) [reply php/libraries/Project.class.inc]: My bad. Don't know why I thought that wasn't the case but I definitely have for a long time.

Thanks

File: test/unittests/NDB_ConfigTest.php (hunk 6 above)
Comment (johnsaigle) [reply test/unittests/NDB_ConfigTest.php]: Just if you're gonna add something new a constant would be cool. not a bit deal though, just relating to our discussion on #5432

File: php/libraries/Project.class.inc (hunk 3 above)
Comment (johnsaigle) [reply php/libraries/Project.class.inc]: In terms of static analysis it's a bit better to have functions that deal strictly with strings if they're meant to be strings instead of `string|null`. If something can be achieved with a stricter type it's usually helpful.

I would flip it around and say what is the advantage of `null` when we can just use an empty string?

Comment (ridz1208) [reply php/libraries/Project.class.inc]: i see you point for the static analyses.

there are 2 things here I wanna make sure not to confuse

 - the nullable `?` is because given the current logic of the class, that would allow people the set the value and then choose the persist it by calling the object function with no arguments. i would hate seeing anywhere is the code `updateName(null)` or `updateName('')`
- the `= null` part, is just because i prefer doing null checks then empty string checks when possible. in this specific case it doesnt make a difference since the empty function returns true to both. So i can change it to `''`

Comment (johnsaigle) [reply php/libraries/Project.class.inc]: Yep that sounds good to me. Like I said I thing `string $projectName = ''` is a bit better to `?string $projectName = null` but in this case specifically they're logically equivalent and either works.

File: php/libraries/Project.class.inc (hunk 5 above)
Comment (ridz1208) [reply php/libraries/Project.class.inc]: because PHP doesnt make sens

$x=null
isset($x) -> flase

:confused:

**Comment by johnsaigle** [comment]
@ridz1208 I did a PHPCBF for you

**Comment by ridz1208** [comment]
@johnsaigle thanks for the phpcbf.

I made the change null to string (phpdoc didnt need to be corrected since the type was string already). I left the `?` for the reasons described above. Does that work ?

**Comment by johnsaigle** [comment]
> the nullable ? is because given the current logic of the class, that would allow people the set the value and then choose the persist it by calling the object function with no arguments. i would hate seeing anywhere is the code updateName(null) or updateName('')

You can still do `updateName()`with no arguments as long as there is a default argument (in this case `''` the empty string). You don't need `?`.

**Comment by ridz1208** [comment]
@johnsaigle thats absolutely true lool

Can u please remove them ?
